types/index.d.ts

```diff
@@ -0,0 +1,127 @@
+interface Array<T> {
+  $remove(item: T): Array<T>;
+  $set(index: any, val: T): T;
+}
+
+type Constructor = {
+  new (...args: any[]): any;
+}
+
+declare class Vue {
+
+  constructor(options?: Vue.ComponentOption);
+
+  $data: Object;
+  readonly $el: HTMLElement;
+  readonly $options: Vue.ComponentOption;
+  readonly $parent: Vue;
+  readonly $root: Vue;
+  readonly $chldren: Vue[];
+  readonly $refs: {[key: string]: Vue};
+  readonly $isServer: boolean;
+
+  $watch(expOrFn: string | Function, callback: Vue.WatchHandler, options: Vue.WatchOption): (() => void);
+  $set: typeof Vue.set;
+  $delete: typeof Vue.delete;
+
+  $on(event: string, callback: Function): this;
+  $once(event: string, callback: Function): this;
+  $off(event?: string, callback?: Function): this;
+  $emit(event: string, ...args: any[]): this;
+
+  $nextTick(callback?: (this: this) => void): void;
+
+  $mount(elementOrSelector?: Element | String, hydrating?: boolean): this;
+  $destroy(): void;
+
+
+  static config: {
+    silent: boolean;
+    optionMergeStrategies: any;
+    devtools: boolean;
+    errorHandler(err: Error, vm: Vue): void;
+    keyCodes: { [key: string]: number };
+  }
+
+  static extend(options: Vue.ComponentOption): Vue;
+  static nextTick(callback: () => void, context?: any[]): void;
+  static set<T>(object: Object, key: string, value: T): T;
+  static delete(object: Object, key: string): void;
+  static directive(id: string, definition?: Vue.DirectiveOption | Vue.DirectiveFunction): Vue.DirectiveOption;
+  static filter(id: string, definition?: Function): Function;
+  static component(id: string, definition?: typeof Vue | Vue.ComponentOption): typeof Vue;
+  static use<T>(plugin: Vue.Plugin<T> | Vue.Installer<T>, options?: T): void;
+  static mixin(mixin: typeof Vue | Vue.ComponentOption): void;
+  static compile(template: string): {render: Function, staticRenderFns: Function};
+}
+
+declare namespace Vue {
```

Comment:
I think we no longer need namespaces for declarations because we can use external module definition format.
Instead of this, I think it would be better to separate files and import the types as needed.